sdks/python/apache_beam/transforms/deduplicate.py

```diff
@@ -0,0 +1,108 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+"""a collection of ptransforms for deduplicating elements."""
+
+from __future__ import absolute_import
+from __future__ import division
+
+from apache_beam.coders.coders import BooleanCoder
+from apache_beam.transforms import ptransform
+from apache_beam.transforms import userstate
+from apache_beam.transforms.core import DoFn
+from apache_beam.transforms.core import Map
+from apache_beam.transforms.core import ParDo
+from apache_beam.transforms.timeutil import TimeDomain
+from apache_beam.typehints import trivial_inference
+from apache_beam.utils.timestamp import Duration
+from apache_beam.utils.timestamp import Timestamp
+
+__all__ = [
+    'DeduplictaionWithinDuration',
+]
+
+
+class DeduplictaionWithinDuration(ptransform.PTransform):
+  """ A PTransform which deduplicates input records over a time domain and
+  threshold. Values in different windows will NOT be considered duplicates of
+  each other. Deduplication is best effort.
+
+  The durations specified may impose memory and/or storage requirements within
+  a runner and care might need to be used to ensure that the deduplication time
+  limit is long enough to remove duplicates but short enough to not cause
+  performance problems within a runner. Each runner may provide an optimized
+  implementation of their choice using the deduplication time domain and
+  threshold specified.
+
+  Does not preserve any order the input PCollection might have had.
+  """
+  def __init__(
+      self, time_domain=TimeDomain.REAL_TIME, duration=Duration(10 * 60)):
+    if time_domain not in (TimeDomain.WATERMARK, TimeDomain.REAL_TIME):
+      raise ValueError(
+          'Unsupported TimeDomain: %r. DeduplictaionWithinDuration'
+          'expects TimeDomain.WATERMARK or TimeDomain.REAL_TIME' %
+          (time_domain, ))
+    self.time_domain = time_domain
+    self.duration = duration
+
+  def _create_deduplicate_fn(self):
+    timer_spec = userstate.TimerSpec('expiry_timer', self.time_domain)
+    state_spec = userstate.BagStateSpec('seen', BooleanCoder())
+    duration = self.duration
+    domain = self.time_domain
+
+    class DeduplicationFn(DoFn):
+      def process(
+          self,
+          element,
+          ts=DoFn.TimestampParam,
+          seen_state=DoFn.StateParam(state_spec),
+          expiry_timer=DoFn.TimerParam(timer_spec)):
+        if True in seen_state.read():
+          return
+
+        if domain == TimeDomain.REAL_TIME:
+          expiry_timer.set(Timestamp.now() + duration)
+        elif domain == TimeDomain.WATERMARK:
+          expiry_timer.set(ts + duration)
```

Comment:
Suppose one has values with timestamps 10, 1, 12, and duration 10. This will not properly deduplicate because the timer will fire at 11 clearing state before 12 gets seen.

Reply:
Are you talking about the watermark case? yes the timer will be fires when the watermark advance to 11 if the timer is set by the value with timestamp 1